app/models/metasploit_data_models/module_run.rb

```diff
@@ -0,0 +1,175 @@
+  #   The name of the protocol that the host was attacked on, if any
+  #   @return [String]
+
+  # @!attribute [rw] session_id
+  #   The {Mdm::Session} that this was run with, in the case of a post module. In exploit modules, this field will
+  #   remain null.
+  #   @return [Datetime]
+
+  # @!attribute [rw] status
+  #   The result of running the module
+  #   @return [String]
+
+  # @!attribute [rw] username
+  #   The name of the user running this module
+  #   @return [String]
+
+
+
+  #
+  # Associations
+  #
+
+
+
+  # @!attribute [rw] loots
+  #  The sweet, sweet loot taken by this module_run
+  #
+  #  @return [ActiveRecord::Relation<Mdm::Loot>]
+  has_many :loots,
+           class_name: 'Mdm::Loot',
+           inverse_of: :module_run
+
+
+  # @!attribute [rw] spawned_session
+  #
+  #  The session created by running this module.
+  #  Note that this is NOT the session that modules are run on.
+  #
+  #  @return [Mdm::Session]
+  has_one :spawned_session,
+             class_name: 'Mdm::Session',
+             inverse_of: :originating_module_run
+
+
+  # @!attribute [rw] target_session
+  #
+  #  The session this module was run on, if any.
+  #  Note that this is NOT a session created by this module run
+  #  of exploit modules.
+  #
+  #  @return [Mdm::Session]
+  belongs_to :target_session,
+             class_name: 'Mdm::Session',
+             foreign_key: :session_id,
+             inverse_of: :target_module_runs
+
+
+
+  # Declares this model to implement a polymorphic relationship with other models.
+  belongs_to :trackable, polymorphic: true
+
+
+  # @!attribute [rw] user
+  #
+  #  The user that launched this module
+  #
+  #  @return [Mdm::User]
+  belongs_to :user,
+             class_name:  'Mdm::User',
+             foreign_key: 'user_id',
+             inverse_of: :module_runs
+
+
+
+  #
+  # Validations
+  #
+
+  # When the module was run
+  validates :attempted_at,
+            presence: true
+
+  # spawned_session is only valid for *exploit modules*
+  validate :no_spawned_session_for_non_exploits
+
+  # target_session is only valid for *non-exploit modules*
+  validate :no_target_session_for_exploits
+
+  # Can't save without information on what module has run
+  validate :module_information_is_present
+
+  # Result of running the module
+  validates :status,
+            inclusion: VALID_STATUSES
+
+
+  private
+
+  # Mark the object as invalid if there is no associated #module_name or {Mdm::ModuleDetail}
+  # @return [void]
+  def module_information_is_present
+    if module_full_name.blank?
+      errors.add(:base, "One of module_name or module_detail_id must be set")
+    end
+  end
+
+  # Mark the object as invalid if there is a spawned_session but the module is *not* an exploit
+  # @return [void]
+  def no_spawned_session_for_non_exploits
+    return true unless spawned_session.present? # nothing to do unless spawned_session is set
+
+    if module_full_name.split('/').first != 'exploit'
```

Comment:
Aux modules can spawn sessions, too.  It may not be important in this case, since (at the moment) the only ones that can, do so via credentials.